### ISSUE TITLE
tealdeer (tldr) does not seem to work properly with ble.sh and a few questions

### ISSUE THREAD

**Issue by Bluey26** [akinomyoga/ble.sh#451]
**ble version**:  `0.4.0-devel4+b72d78a9`
**Bash version**:  `5.2.26(1)-release (x86_64-pc-linux-gnu)`

Hello, i started to use ble.sh because it allwos me to see my history commands without open the history file. I removed some features which i was not interested to use but i still struggle with a few things:

I use a command called `tldr` (https://github.com/tldr-pages/tldr/) which shows concise information about how to use a program, for example: `tldr pacman` will show the basic and most used commands for the pacman package manager.

This command works properly in pure bash and in zsh for example, but when i installed ble.sh (`0.4.0-devel4+b72d78a9`) i end up with a bug or some kind of glitch that i dont know how to solve. I will attach a short video showing what is happening.

https://github.com/akinomyoga/ble.sh/assets/143142826/72cd4349-4e69-44e4-aa54-44da01eee9d4

It happens to me with my default terminal (qterminal) but it also happens in Xterm.

I suspect this is related to some kind of vim or the sort hotkey, which i would like to remove if you can tell me how (to avoid conflicts with normal bash usage).

Keep in mind that the shown message that is repeated only is shown by tldr after the command is executed (is some kind of warning) and is shown only one time per tldr execution.

---

- In the other hand, i would like to remove the ` ble: something` status messages that appear in the terminal from time to time. I think this is not related to the main issue of this ticket but i don't know how to do it.

- I also dont know how to do the following: i like the feature that shows the history-based commands, but i would like to only keep that part, and not the 'smart' part that says me the name of a file starting with the letters i am writing.

For example: I write: `ffmpeg -i d` and i get a suggestion in my $HOME dir with a file which starts with a "d". I dont know how this is called and if its possible, but sometimes interferes with the history-based suggestions (that i like to keep). I only would like to keep the 'i write `ffmpeg -i d` and when i press TAB i will see any possible file that starts wid the "d".

I really appreciate the program you have made, i disabled most functions because i dont use them but please dont take it personal, i imagine a lot of people find them useful.

**PS:** If the video does not play, download it. For some reason i am unable to see it in firefox, i had to download it. 
The video shows a continious "spam" of the tldr warning. This message appears when i am writing letters.

**PS2:** The `tldr pacman` command and every other saved in the history shows properly. It only happens for those i dont save. If i run `tldr cat` ,even with the glitch, and then i open a new terminal and i write` tldr cat`, the issue does not happen with that command

**Comment by akinomyoga** [comment]
## tldr cache issue

> I use a command called `tldr` (https://github.com/tldr-pages/tldr/) which shows concise information about how to use a program, for example: `tldr pacman` will show the basic and most used commands for the pacman package manager.
> 
> This command works properly in pure bash and in zsh for example, but when i installed ble.sh (`0.4.0-devel4+b72d78a9`) i end up with a bug or some kind of glitch that i dont know how to solve. I will attach a short video showing what is happening.

Thanks for the report. I tried to reproduce the problem, but so far I cannot reproduce the problem. The `tldr` command seems to just work without problems within ble.sh sessions.

- **Q1**: Does the problem reproduce with a plain Bash plus just ble.sh (without any other custom configuration)? For example, could you check if the problem arises within a session started with the following way?

```bash
$ cp ~/.bash_history ~/.bash_history.backup # please back up your command history if needed
$ INPUTRC=/dev/null bash --norc --noprofile # <-- start a child session of plain Bash without any configuration
$ source /path/to/ble.sh --norc
$ tldr xxxx
# <-- Does the problem reproduce?
```

> I suspect this is related to some kind of vim or the sort hotkey, which i would like to remove if you can tell me how (to avoid conflicts with normal bash usage).

- **Q2**: Could you explain it in more detail? What is "the vim or the sort hotkey"? As far as I see the video you attached, you don't seem to use the vi-editing mode. Or do you mean you have additional settings for Vim that may interfere with Bash?

I actually suspect this would be related to the problem with `tldr` that can be triggered by some factors.

Another mystery is that as far as I search the shown error message, [nothing is found in the entire project of tldr on GitHub](https://github.com/search?q=org%3Atldr-pages%20%22The%20cache%20hasn%27t%20been%20updated%20for%22&type=code). In addition, [nothing seems to be found on the internet](https://www.google.co.jp/search?q=%22The+cache+hasn%27t+been+updated+for%22).

###

> * In the other hand, i would like to remove the ` ble: something` status messages that appear in the terminal from time to time. I think this is not related to the main issue of this ticket but i don't know how to do it.

Have you checked [README §2.2](https://github.com/akinomyoga/ble.sh?tab=readme-ov-file#22-disable-features)? In particular, have you tried the last item?

```bash
# Disable some other markers like "[ble: ...]"
bleopt edit_marker=
bleopt edit_marker_error=
```

The detailed explanations are found in [Manual §4.15.18](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A74-Editing#user-content-bleopt-edit_marker) and [Maunal §4.15.19](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A74-Editing#user-content-bleopt-edit_marker_error).

### Disabling autosuggestion based on syntax and programmable completions

> * I also dont know how to do the following: i like the feature that shows the history-based commands, but i would like to only keep that part, and not the 'smart' part that says me the name of a file starting with the letters i am writing.

```bash
# blerc

ble-import lib/core-complete -C 'ble/array#remove _ble_complete_auto_source syntax'
```

> I really appreciate the program you have made, i disabled most functions because i dont use them but please dont take it personal, i imagine a lot of people find them useful.

Thanks.

**Comment by akinomyoga** [comment]
## Re: tldr cache issue

> I use a command called `tldr` (https://github.com/tldr-pages/tldr/) which shows concise information about how to use a program, for example: `tldr pacman` will show the basic and most used commands for the pacman package manager.

I suspect what you use is not https://github.com/tldr-pages/tldr/ but https://github.com/dbrgn/tealdeer because I found the error message in the video in the codebase of `tealdeer`.

However, I still cannot reproduce the problem with the tealdeer version of `tldr`. I modified `updates.auto_update_interval_hours` in `config.toml` to shorten the expiration of the cache, and I could get the message 

```
Warning: The cache hasn't been updated for 0 days.
You should probably run `tldr --update` soon.
```

but the error message is not repeated and the command execution just terminates without any problems. Another thing that I noticed is that the error message is still slightly different; there is an extra word "Warning: " at the beginning in my environment. Maybe the version of `tealdeer` is different:

Could you provide the version of your `tldr`? What is the result of the following command?

```bash
$ tldr --version
```

> It happens to me with my default terminal (qterminal) but it also happens in Xterm.

I tried it in my terminal, xterm, and screen, but the problem doesn't arise in any terminals in my environment.

**Comment by Bluey26** [comment]
Hello @akinomyoga 

First of all thank you for you detailed explanations.
Yes, i made a mistake, i am not using `tldr` but `tealdeer`. It just that its been a while since i installed it, and its triggered by using` tldr`

**Q1**

> $ cp ~/.bash_history ~/.bash_history.backup # please back up your command history if needed
> $ INPUTRC=/dev/null bash --norc --noprofile # <-- start a child session of plain Bash without any configuration
> $ source /path/to/ble.sh --norc
> $ tldr xxxx
>  <-- Does the problem reproduce?

The problem does NOT reproduce with this steps. This will likely indicate that the problem is, around my configuration. I modified too much the ble.sh default configs. I will detail in this in a few lines.


**Q2:**

I was suspecting some kind of binding. I thought of vim-like because i have read in some issues that those are somewhat integrated in ble.sh , but i think i may be wrong about this, since you said they are in vi-editing mode.


**Hiding edit markers of the form...**

Yes, i have those enabled (will show them in a few lines) but there is still some around, for example i recall one saying `[error 1]` (running `ffmpeg` will show it)

**Disabling autosuggestion based on syntax and programmable completions**

Yes, this seems what i wanted, thank you!!!!

> Could you provide the version of your tldr?

`tealdeer 1.6.1`

I was doing some tests meanwhile doing this comment, and i found that the problem would be likely in my `.bashrc` (defaulting the `.blerc` produced no changes). I say this because i removed temporarily the lines present in my .bashrc and then opened a terminal tab. Everything was OK, the default ble.sh configuration. Then, again, i pasted my ble.sh lines and now the error is not present anymore (strange).
In any case, i will add the ble.sh-related lines from my .bashrc in here, because the problem could arise in any moment.


```
# Blesh Start (line 7 of my bashrc)
[[ $- == *i* ]] && source /usr/share/blesh/ble.sh --noattach

# Disable syntax highlighting
bleopt highlight_syntax=

# Disable highlighting based on filenames
bleopt highlight_filename=

# Disable highlighting based on variable types
bleopt highlight_variable=



# Disable auto-complete (Note: auto-complete is enabled by default in bash-4.0+)
#bleopt complete_auto_complete=
# Tip: you may instead specify the delay of auto-complete in millisecond
#bleopt complete_auto_delay=300

# Disable auto-complete based on the command history
#bleopt complete_auto_history=



# Disable ambiguous completion
bleopt complete_ambiguous=

# Disable menu-complete by TAB
#bleopt complete_menu_complete=

# Disable menu filtering (Note: auto-complete is enabled by default in bash-4.0+)
bleopt complete_menu_filter=

# Disable EOF marker like "[ble: EOF]"
bleopt prompt_eol_mark=''
# Tip: you may instead specify another string:
#bleopt prompt_eol_mark='⏎'

# Disable error exit marker like "[ble: exit %d]"
bleopt exec_errexit_mark=
# Tip: you may instead specify another string:
bleopt exec_errexit_mark=$'\e[91m[error %d]\e[m'

# Disable elapsed-time marker like "[ble: elapsed 1.203s (CPU 0.4%)]"
bleopt exec_elapsed_mark=
# Tip: you may instead specify another string
#bleopt exec_elapsed_mark=$'\e[94m[%ss (%s %%)]\e[m'
# Tip: you may instead change the threshold of showing the mark
#bleopt exec_elapsed_enabled='sys+usr>=10*60*1000' # e.g. ten minutes for total CPU usage

# Disable exit marker like "[ble: exit]"
bleopt exec_exit_mark=

# Disable some other markers like "[ble: ...]"
bleopt edit_marker=
bleopt edit_marker_error=

# CTRL+Backspace delete word for blesh
ble-bind -f '^?' kill-backward-cword

[[ ${BLE_VERSION-} ]] && ble-attach 
#The last line of my .bashrc
```

As you can see, i adapted the [Disable Features](https://github.com/akinomyoga/ble.sh?tab=readme-ov-file#22-disable-features) section and i was doing some commenting in there to see what i like to keep.

Another small detail: Before this (a few hours ago) i discovered that if i wrote `tldr -q something` the error did not appear(but it appeared with `tldr something`) so this could indicate some kind of 'cache' error, but i am new in all this, just more information that could help.

**Comment by akinomyoga** [comment]
Thank you for your replies.

> **Hiding edit markers of the form...**
> 
> [...], so this part is solved.

OK!

> > Could you provide the version of your tldr?
> 
> `tealdeer 1.6.1`

This seems to be the same as mine, which I built from the source today.

> i found that the problem would be likely in my `.bashrc` (defaulting the `.blerc` produced no changes). I say this because i removed temporarily the lines present in my .bashrc (i removed the customizations, not the startup and end lines to make ble.sh work) and then opened a terminal tab. Everything was OK, the default ble.sh configuration. Then, again, i pasted my ble.sh lines and now the error is not present anymore (strange).

Hmm, if the error is not present when you restore the ble.sh settings, that suggests that those ble.sh settings are unrelated. I also think these shouldn't affect the behavior of external programs. In fact, I tried your ble.sh settings now, but the problem doesn't arise with `tldr`.

I suspect you have changed another part of your system since you last reproduced the problem. For example, did you run `tldr --update`? If so, I guess you'll need to wait for 30 days to reproduce the problem again... because `tldr` doesn't try to trigger the feature for 30 days after the last update. Or for a testing purpose, you can modify the source code of `tealdeer` as e.g.

```diff
diff --git a/src/config.rs b/src/config.rs
index 98d4e46..7cd896f 100644
--- a/src/config.rs
+++ b/src/config.rs
@@ -14,7 +14,7 @@ use yansi::{Color, Style};
 use crate::types::PathSource;

 pub const CONFIG_FILE_NAME: &str = "config.toml";
-pub const MAX_CACHE_AGE: Duration = Duration::from_secs(2_592_000); // 30 days
+pub const MAX_CACHE_AGE: Duration = Duration::from_secs(60); // 1 minute
 const DEFAULT_UPDATE_INTERVAL_HOURS: u64 = MAX_CACHE_AGE.as_secs() / 3600; // 30 days

 fn default_underline() -> bool {
```

and install this modified version.

**Comment by Bluey26** [comment]
No, strangely it works all of sudden.

I have not updated the cache, just because it would make this not reproducible anymore:

```
> tldr cat
Warning: The cache hasn't been updated for 42 days.
You should probably run `tldr --update` soon.     
```

**Comment by akinomyoga** [comment]
Thank you for keeping your cache for the testing purpose. So there should be any other part that has been changed. One possibility is that the ble.sh settings you had were actually different from the current one, but I think there are even other factors that might affect the behavior.

I'd like to ask about the detailed behavior in the video. I see some stray characters are inserted at the beginning of the error messages, which reads

```
co
oo
oo
oo
op
pc
cx
xz
za
as
sd
dj
```

Do you have an idea about what these are? Maybe are they what you've input through your keyboard? I've been thinking until now that those are random strings produced by a single call of `tldr`, and those many lines are produced automatically without any users' inputs, but maybe the error message was output every time you press a key?

I was also wondering why the header `Warning: ` was missing when the error happened, but I now realize that `tldr` command puts `Warning: ` only when stdout is connected to TTY.

```
$ tldr fdsafdsa 2> err.txt
$ cat err.txt
Warning: The cache hasn't been updated for 40 days.
You should probably run `tldr --update` soon.
Warning: Page `fdsafdsa` not found in cache.
Try updating with `tldr --update`, or submit a pull request to:
https://github.com/tldr-pages/tldr

$ tldr fdsafdsa 2> err.txt >/dev/null
$ cat err.txt
The cache hasn't been updated for 40 days.
You should probably run `tldr --update` soon.
Page `fdsafdsa` not found in cache.
Try updating with `tldr --update`, or submit a pull request to:
https://github.com/tldr-pages/tldr
```

This implies that when the error is happening, `stdout` of tldr is not connected to stdout.

Note: the above message says *40 days*, but I'm not testing the possibility that the problem only happens at a specific day count by modifying the source code of tealdear. However, I still cannot reproduce the problem.

**Comment by akinomyoga** [comment]
> maybe the error message was output every time you press a key?

Ah, wait. I think I now understand what happened. So in the video, you haven't actually run the `tldr` command by hitting <kbd>Enter</kbd>, but were you just inputting the string containing `tldr xxxx`? In the video form of reports, I can only see the result and cannot see what you've done at all, which is a problem...

If that's the case, it should be related to a completion setting for the `tldr` command. This is consistent with the fact that the `tldr` error message is in the form for non-TTY stdout because ble.sh suppresses stdout of completion settings. Then I found a completion setting for Bash in the repository of tealdear [here](https://github.com/dbrgn/tealdeer/blob/main/completion/bash_tealdeer), which seems to assume `bash-completion`.

Do you have a file `/usr/share/bash-completion/completions/tldr` in your system?

I guess you had the code to load bash-completion in the previous setting (such as `source /usr/share/bash-completion/bash_completion` or `source /etc/profile.d/bash_completion` or `source /etc/bashrc` or `source /etc/bash.bashrc`, etc. which depends on the Linux distribution), and the `tealdear` completion was loaded by `bash-completion`. If so, I think you can reproduce the problem by including `source /usr/share/bash-completion/bash_completion` in your `~/.bashrc` as well as `source /path/to/ble.sh`.

**Comment by Bluey26** [comment]
> ... Do you have an idea about what these are? Maybe are they what you've input through your keyboard?...

Those are my keyboard inputs. When i started to typing, the warning message about the cache update started to appear, 1 time per letter. The keys i press were random, they are not an actual valid string (like `cat` `pacman` or `echo`) and were just to spam the warning message.

> ...  maybe the error message was output every time you press a key?

Indeed, that was what happened, but the strangest part was that the error was not reproduced if the string was already stored in the history file (the first `pacman` command did not triggered it, because as its shown by the gray suggestion, it was already in the history file). This happened for every string in the history file, so if for example i wrote `tldr wallpaper`, the first time i would see all those warnings per each keyboard input key i used. Then, if i pressed enter, the command would run properly and would be stored in my .bash_history file. If i close that terminal and open a new terminal window, the issue would not happen with the command `tldr wallpaper` because it was stored in the history.

> ... but I'm not testing the possibility that the problem only happens at a specific day count by modifying the source code of tealdear. However, I still cannot reproduce the problem.

I dont think that would be the case. Because i reported this a 3 days ago and it was happening to me just 2 hours ago (during the answer to your reply). So it should be day-independent.

> tldr command puts Warning: only when stdout is connected to TTY.

In other words, because the ENTER key was not pressed, right?

Could be all this be caused by some kind of incomplete removal? Because of the following: I installed the [blesh](https://aur.archlinux.org/packages/blesh) package first, and then i discovered that some functions are not in the 0.3 and that it would be easier to me to use the [blesh-git](https://aur.archlinux.org/packages/blesh-git) package instead (the ble-faces were easier with git since all issues were using the devel version).

What i do not recall was if i removed it and then installed the `-git` package or if i ran `pacman -U ...-git.tar.xz`

Another thing was that the first time of using `blesh-git` package was using the 0.3 source command:

`[[ $- == *i* ]] && source /usr/share/blesh/ble.sh`

But this should have affected me just a few hours, since i changed that line for the devel suggested ones (present in a previous comment, in where i added my ble-sh ` .bashrc` lines

---

About your new comment, i have the `bash_completion` file in `/usr/share/bash-completion/` but it was not in my `.bashrc` . In the other hand, adding it or sourcing it inside a new terminal tab does not reproduce the issue.

I also have `/usr/share/bash-completion/completions/tldr`

**Comment by akinomyoga** [comment]
I tried the current version of [the tealdear completion](https://github.com/dbrgn/tealdeer/blob/main/completion/bash_tealdeer), but the problem doesn't reproduce. I also checked older versions of the `tldr` completions (found in the following lists) but could not find suspicious changes.

- [History for completion/bash_tealdeer - dbrgn/tealdeer](https://github.com/dbrgn/tealdeer/commits/main/completion/bash_tealdeer)
- [History for bash_tealdeer - dbrgn/tealdeer](https://github.com/dbrgn/tealdeer/commits/21772978718a71e9d216cdcae2431bf08853a843/bash_tealdeer?browsing_rename_history=true&amp;new_path=completion/bash_tealdeer&amp;original_branch=main)

Then, I next suspected that a completion setting for a different `tldr` implementation was activated. I found the completion setting for [`tldr-pages/tldr`](https://github.com/tldr-pages/tldr/) at [`tldr-node-client/bin/completion/bash/tldr`](https://github.com/tldr-pages/tldr-node-client/blob/master/bin/completion/bash/tldr). I also tried this and this produces error messages, though the content of the error message is different from yours.

![image](https://github.com/akinomyoga/ble.sh/assets/8982192/cc2faadb-025f-480b-9f61-9ead01c36f0e)

I guess you had a completion setting for a different `tldr` implementation, though I'm not sure which one was the target implementation of the `tldr` completion you had.

----

> > ... Do you have an idea about what these are? Maybe are they what you've input through your keyboard?...
> 
> Those are my keyboard inputs.

Thanks for the confirmation!

> > ...  maybe the error message was output every time you press a key?
> 
> Indeed, that was what happened, but the strangest part was that the error was not reproduced if the string was already stored in the history file

This is also consistent with the current hypothesis that the completion settings had caused the problem. The auto-complete feature is involved here. The `auto-complete` feature is attempted every time you insert a new character in the command line. `auto-complete` first attempts the history completion, and when it fails, it next attempts the programmable completion based on the user's completion setting. When you don't have a matching history entry, the first history completion fails, so ble.sh next attempts the user's completion setting for `tldr`. When you have already a matching history entry, the first history completion succeeds, and the later attempt for the user's completion settings is skipped.

Also, do you still have the line `ble-import lib/core-complete -C 'ble/array#remove _ble_complete_auto_source syntax'` in your `~/.blerc` while testing this? If so, that setting also suppresses the problem because that setting disables the second attempt of the user's completion setting.

> > tldr command puts Warning: only when stdout is connected to TTY.
> 
> In other words, because the ENTER key was not pressed, right?

Ah, yes. I was thinking you've run the command by pressing ENTER key, so I was trying to reproduce the problem by pressing ENTER key, where stdout is connected to TTY and the form of the error message is different.

> Could be all this be caused by some kind of incomplete removal? Because of the following: I installed the [blesh](https://aur.archlinux.org/packages/blesh) package first, and then i discovered that some functions are not in the 0.3 and that it would be easier to me to use the [blesh-git](https://aur.archlinux.org/packages/blesh-git) package instead (the ble-faces were easier with git since all issues were using the devel version).

The incomplete removal of `ble.sh` shouldn't be related. If you had a different `tldr` implementation in your system in the past, that could be related. The possibility I now suspect is that the completion setting in other `tldr` implementation remained in your system or in your Bash settings, which caused the problem.

> What i do not recall was if i removed it and then installed the `-git` package or if i ran `pacman -U ...-git.tar.xz`

You don't have to remove the previous version (unless the used AUR helper is broken). [blesh-git/PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=blesh-git#n25) specifies `provides` and `conflicts`, so installing `blesh-git` should automatically remove the other version of  `blesh` package before installing `blesh-git`.
 
> Another thing was that the first time of using `blesh-git` package was using the 0.3 source command:
> 
> `[[ $- == *i* ]] && source /usr/share/blesh/ble.sh`
> 
> But this should have affected me just a few hours, since i changed that line for the devel suggested ones (present in a previous comment, in where i added my ble-sh ` .bashrc` lines
> 
> About your new comment, i have the `bash_completion` file in `/usr/share/bash-completion/` but it was not in my `.bashrc` . In the other hand, adding it or sourcing it inside a new terminal tab does not reproduce the issue.

I now think it should also be related to the line `ble-import lib/core-complete -C 'ble/array#remove _ble_complete_auto_source syntax'`, which I guess you have now in `~/.blerc`.

> I also have `/usr/share/bash-completion/completions/tldr`

Could you provide the content of that file?

Also what is the result of the following command?

```bash
$ complete -p tldr
```

**Comment by Bluey26** [comment]
> Could you provide the content of that file?

Yes, of course:

[tldr-completions-bash.txt](https://github.com/akinomyoga/ble.sh/files/15469934/tldr-completions-bash.txt)


`complete -p tldr` shows:

` bash: complete: tldr: no completion specification`


`complete -F _tealdeer tldr` (this one is shown in the file i attached) shows no output.

I instaled ` tealdeer` using pacman (its on the arch official repositories) and i dont remember to have install anything related using `make install` or the sort.


> I now think it should also be related to the line ble-import lib/core-complete -C 'ble/array#remove _ble_complete_auto_source syntax', which I guess you have now in ~/.blerc.

Indeed, commenting that line in my `.blerc` and running `source /usr/share/bash-completion/bash_completion` does reproduce the issue with my config files. Now every time i open a new terminal window and try `tldr something` the problem arises again, so this line _bypass_ the issue, somehow.

**Comment by Bluey26** [comment]
Digging a bit deeper, i have found that the `tealdeer` package was updated 1 year ago [archlinux-archive](https://archive.archlinux.org/packages/t/tealdeer/)

If i do the `diff` between the file provided by` tealdeer-1.6.1-1-x86_64.pkg.tar.zst` and my file, i dont see differences(no output).

So this could be some kind of issue with tealdeer, indeed

**Comment by akinomyoga** [comment]
OK, the minimal reproducer of the error message is this:

```bash
$ tldr -l >/dev/null
The cache hasn't been updated for 40 days.
You should probably run `tldr --update` soon.
```

The up-to-date version of the completion setting suppresses the error message by specifying `2>/dev/null`.

**Comment by akinomyoga** [comment]
I found the corresponding PR in the upstream tealdeer: https://github.com/dbrgn/tealdeer/pull/327, though the content of the error message seems to have been different from ours. The reporter had experienced the error message for a different reason, i.e., the missing cache.

**Comment by Bluey26** [comment]
I think this one could be closed, since as you have tagged, its a external issue. Thank you very much for you help!!!

**Comment by akinomyoga** [comment]
I'm closing. Thanks!
